cub/benchmarks/bench/scan/applications/rabin-karp-second-fingerprinting.cu

```diff
@@ -0,0 +1,372 @@
+
+// Types associated with the cyclic ring
+using ZpT   = cuda::std::uint32_t;
+using WideT = cuda::std::uint64_t;
+
+using MatT = cuda::std::array<ZpT, 4>;
+
+inline ZpT __host__ __device__ Zp_mul(ZpT v1, ZpT v2, cuda::fast_mod_div<WideT> m_p)
+{
+  const auto w1 = static_cast<WideT>(v1);
+  const auto w2 = static_cast<WideT>(v2);
+  return static_cast<ZpT>((w1 * w2) % m_p);
+}
+
+inline ZpT __host__ __device__ Zp_add(ZpT v1, ZpT v2, cuda::fast_mod_div<WideT> m_p)
+{
+  const auto w1 = static_cast<WideT>(v1);
+  const auto w2 = static_cast<WideT>(v2);
+  return static_cast<ZpT>((w1 + w2) % m_p);
+}
+
+inline MatT __host__ __device__ Zp_matmul(MatT v1, MatT v2, cuda::fast_mod_div<WideT> m_p)
+{
+  ZpT _1_00_2_00 = Zp_mul(v1[0], v2[0], m_p);
+  ZpT _1_01_2_10 = Zp_mul(v1[1], v2[2], m_p);
+  ZpT _r_00      = Zp_add(_1_00_2_00, _1_01_2_10, m_p);
+
+  ZpT _1_00_2_01 = Zp_mul(v1[0], v2[1], m_p);
+  ZpT _1_01_2_11 = Zp_mul(v1[1], v2[3], m_p);
+  ZpT _r_01      = Zp_add(_1_00_2_01, _1_01_2_11, m_p);
+
+  ZpT _1_10_2_00 = Zp_mul(v1[2], v2[0], m_p);
+  ZpT _1_11_2_10 = Zp_mul(v1[3], v2[2], m_p);
+  ZpT _r_10      = Zp_add(_1_10_2_00, _1_11_2_10, m_p);
+
+  ZpT _1_10_2_01 = Zp_mul(v1[2], v2[1], m_p);
+  ZpT _1_11_2_11 = Zp_mul(v1[3], v2[3], m_p);
+  ZpT _r_11      = Zp_add(_1_10_2_01, _1_11_2_11, m_p);
+
+  return {_r_00, _r_01, _r_10, _r_11};
+}
+
+struct RabinKarpOp
+{
+  cuda::fast_mod_div<WideT> m_p;
+
+  __host__ __device__ RabinKarpOp(ZpT p)
+      : m_p(static_cast<WideT>(p))
+  {}
+
+  MatT __host__ __device__ operator()(MatT v1, MatT v2) const
+  {
+    return Zp_matmul(v1, v2, m_p);
+  }
+};
+
+template <typename T>
+struct ChunkToMat
+{
+  static_assert(cuda::std::is_integral_v<T> && cuda::std::is_unsigned_v<T>, "Bit sequence should be represented");
+
+  cuda::fast_mod_div<WideT> m_p;
+
+  __host__ __device__ ChunkToMat(ZpT p)
+      : m_p(static_cast<WideT>(p))
+  {}
+
+  MatT __host__ __device__ operator()(const T& bits) const
+  {
+    static constexpr int n_bits = cuda::std::numeric_limits<T>::digits;
+    static constexpr MatT _id   = {ZpT{1}, ZpT{0}, ZpT{0}, ZpT{1}}; // [[1, 0], [0, 1]]
+    static constexpr MatT _0    = {ZpT{1}, ZpT{0}, ZpT{1}, ZpT{1}}; // [[1, 0], [1, 1]]
+    static constexpr MatT _1    = {ZpT{1}, ZpT{1}, ZpT{0}, ZpT{1}}; // [[1, 1], [0, 1]]
+
+    // initialize with identity matrix
+    MatT m  = _id;
+    T _bits = bits;
+
+    if constexpr (false)
+    {
+      cuda::static_for<n_bits>([&](int) {
+        m = Zp_matmul((_bits & 1) ? _1 : _0, m, m_p);
+        _bits >>= 1;
+      });
+    }
+    else
+    {
+      for (int i = 0; i < n_bits; ++i)
+      {
+        (void) i;
+        m = Zp_matmul((_bits & 1) ? _1 : _0, m, m_p);
+        _bits >>= 1;
+      }
+    }
+
+    return m;
+  }
+};
+
+/*! @brief Iterator that performs assignment at specific index only, discards otherwise
+
+    This iterator allows tp use inclusive_scan to perform reduction with non-commutative,
+    but associative binary operator
+  */
```

Comment:
Suggestion: we don't use Doxygen comments in the CUB benchmarks, maybe replace with a normal comment